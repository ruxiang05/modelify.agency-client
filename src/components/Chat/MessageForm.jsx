/* Uses methods and/or components from react,prop-types and socket.io-client */
import React from 'react';
import PropTypes from 'prop-types';
import { ReactComponent as SendIcon } from '../../assets/icons/arrow-circle-right.svg';

class MessageForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      message: '',
    };

    this.sendMessage = this.sendMessage.bind(this);
    this.handleChange = this.handleChange.bind(this);
  }

  handleChange(event) {
    this.setState({ message: event.target.value });
  }

  sendMessage(event) {
    const { socket, user, chat } = this.props;
    const { message } = this.state;
    event.preventDefault();
    if (!message) return;
    socket.emit('send_message', {
      room: chat._id,
      user: user._id,
      message,
    });

    this.setState({ message: '' });
  }

  render() {
    const { message } = this.state;
    return (
      <div className="message-form">
        <input value={message} onChange={this.handleChange} />
        <button type="submit" onClick={this.sendMessage}><SendIcon /></button>
      </div>
    );
  }
}

MessageForm.propTypes = {
  user: PropTypes.shape({}).isRequired,
  socket: PropTypes.shape({}).isRequired,
  chat: PropTypes.shape({}).isRequired,
};

export default MessageForm;
