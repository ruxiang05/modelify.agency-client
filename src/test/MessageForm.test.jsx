import React from 'react';
import socket from 'socket.io-client';
import { shallow } from 'enzyme';
import MessageForm from '../components/Chat/MessageForm';

let wrapper;

const props = {
  socket: {
    emit: jest.fn(),
  },
  user: {
    _id: '1',
  },
  chat: {
    _id: '2',
  },
};

beforeEach(() => {
  wrapper = shallow(<MessageForm {...props} />);
});

describe('MessageForm', () => {
  it('should render MessageForm', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should handle input change', async () => {
    const event = {
      target: {
        value: 'hi',
      },
    };
    wrapper.instance().handleChange(event);
    await wrapper.update();
    expect(wrapper.state().message).toBe('hi');
  });

  it('should send message', async () => {
    const event = {
      preventDefault: jest.fn(),
    };
    wrapper.setState({ message: 'hi' });
    await wrapper.update();
    wrapper.instance().sendMessage(event);
    expect(props.socket.emit).toHaveBeenCalled();
    expect(wrapper.state().message).toBe('');
  });

  it('should not send message if it is empty', () => {
    socket.emit = jest.fn();
    const event = {
      preventDefault: jest.fn(),
    };
    wrapper.instance().sendMessage(event);
    expect(socket.emit).not.toHaveBeenCalled();
  });
});
