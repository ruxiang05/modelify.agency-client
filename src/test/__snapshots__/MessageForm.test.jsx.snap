// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageForm should render MessageForm 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MessageForm
    chat={
      Object {
        "_id": "2",
      }
    }
    socket={
      Object {
        "emit": [MockFunction],
      }
    }
    user={
      Object {
        "_id": "1",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <input
          onChange={[Function]}
          value=""
        />,
        <button
          onClick={[Function]}
          type="submit"
        >
          <ForwardRef />
        </button>,
      ],
      "className": "message-form",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "onChange": [Function],
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <ForwardRef />,
          "onClick": [Function],
          "type": "submit",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          },
        },
        "type": "button",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            onChange={[Function]}
            value=""
          />,
          <button
            onClick={[Function]}
            type="submit"
          >
            <ForwardRef />
          </button>,
        ],
        "className": "message-form",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "onChange": [Function],
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <ForwardRef />,
            "onClick": [Function],
            "type": "submit",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
          },
          "type": "button",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
